### ISSUE TITLE
Brain tumor dataset dataloader

### ISSUE THREAD

**Issue by kuixing233** [dogeONE-bit/RIOG#1]
Thanks for your wonderful achievements. 

I would like to ask if there is the dataloader for Brain tumor dataset. 

Thanks again.

**Comment by dogeONE-bit** [comment]
Thank you for your interest in our work. The code for our brain tumor dataloader is provided below.  You may copy these directly, or you can leave your email address and I will send the files to you by email. 
import os
import numpy as np
from torch.utils.data import Dataset
import torchvision.transforms as transforms
import SimpleITK as sitk
from PIL import Image
import torch
import glob
import random
import cv2


def adjust_light(img):
    image = img
    seed = random.random()
    if seed > 0.5:
        gamma = random.random() * 3 + 0.5
        invGamma = 1.0 / gamma
        image = np.array(image).astype(np.float32) / 255.0
        image = np.power(image, invGamma)
        image = (image * 255).astype(np.uint8)
        image = np.clip(image, 0, 255)
    return image.astype(np.float32)


def add_salt_and_pepper_noise(image, salt_prob=0.1, pepper_prob=0.1):
    noisy_image = image.copy()
    seed = random.random()
    if seed > 0.75:
        salt_mask = np.random.rand(*image.shape) < salt_prob
        noisy_image[salt_mask] = 255
    elif seed > 0.5:
        pepper_mask = np.random.rand(*image.shape) < pepper_prob
        noisy_image[pepper_mask] = 0
    return noisy_image


def eraser(img):
    s_l = 0.02
    s_h = 0.06
    r_1 = 0.3
    r_2 = 0.6
    v_l = 0
    v_h = 255
    pixel_level = False

    image = img
    img_h, img_w, img_c = image.shape

    if random.random() > 0.5:
        return image

    while True:
        s = np.random.uniform(s_l, s_h) * img_h * img_w
        r = np.random.uniform(r_1, r_2)
        w = int(np.sqrt(s / r))
        h = int(np.sqrt(s * r))
        left = np.random.randint(0, img_w)
        top = np.random.randint(0, img_h)
        if left + w <= img_w and top + h <= img_h:
            break

    if pixel_level:
        c = np.random.uniform(v_l, v_h, (h, w, img_c))
    else:
        c = np.random.uniform(v_l, v_h)

    image[top:top + h, left:left + w, :] = c
    return image


def add_gaussian_noise(image, mean=0, std=25):
    if random.random() > 0.5:
        return image
    noise = np.random.normal(mean, std, image.shape)
    noisy_image = image + noise
    noisy_image = np.clip(noisy_image, 0, 255)
    return noisy_image


def read_img(path):
    img = sitk.ReadImage(path)
    data = sitk.GetArrayFromImage(img)
    return data


def crop_center(img, cropx, cropy):
    _, y, x = img.shape
    startx = x // 2 - (cropx // 2)
    starty = y // 2 - (cropy // 2)
    return img[:, starty:starty + cropy, startx:startx + cropx]


def normalize(img):
    mask = img != 0
    non_zero_img = img[mask]
    mean = np.mean(non_zero_img)
    std = np.std(non_zero_img)
    img[mask] = (non_zero_img - mean) / std
    return img


def normalize2047(x):
    x = x / 2047
    x[x > 1.0] = 1.0
    x = 2 * x - 1
    return x


def reshape(img1, img2, img3, img4, label):
    output_vol_size = 160
    tempL = np.nonzero(img2)

    min_h, max_h = np.min(tempL[1]), np.max(tempL[1])
    min_w, max_w = np.min(tempL[2]), np.max(tempL[2])

    p_h = max(output_vol_size - (max_h - min_h), 0) // 2
    p_w = max(output_vol_size - (max_w - min_w), 0) // 2

    min_h = max(min_h - p_h, 0)
    max_h = min_h + output_vol_size
    min_w = max(min_w - p_w, 0)
    max_w = min_w + output_vol_size

    img1 = img1[:, min_h:max_h, min_w:max_w]
    img2 = img2[:, min_h:max_h, min_w:max_w]
    img3 = img3[:, min_h:max_h, min_w:max_w]
    img4 = img4[:, min_h:max_h, min_w:max_w]
    label = label[:, min_h:max_h, min_w:max_w]

    return img1, img2, img3, img4, label


def reshape_isles(img1, img2, label):
    output_vol_size = [96, 96]
    tempL = np.nonzero(img2)

    min_h, max_h = np.min(tempL[0]), np.max(tempL[0])
    min_w, max_w = np.min(tempL[1]), np.max(tempL[1])

    p_h = max(output_vol_size[0] - (max_h - min_h), 0) // 2
    p_w = max(output_vol_size[1] - (max_w - min_w), 0) // 2

    min_h = max(min_h - p_h, 0)
    max_h = min_h + 96
    min_w = max(min_w - p_w, 0)
    max_w = min_w + 96

    img1 = img1[min_h:max_h, min_w:max_w]
    img2 = img2[min_h:max_h, min_w:max_w]
    label = label[min_h:max_h, min_w:max_w]

    return img1, img2, label


def re_label(label):
    _, H, W = label.shape
    outputs = np.zeros((3, H, W))
    outputs[0, :, :] = np.where(label != 0, 1, 0)
    outputs[1, :, :] = np.where((label == 1) | (label == 3), 1, 0)
    outputs[2, :, :] = np.where(label == 3, 1, 0)
    return outputs


def add_gibbs_noise(data, probability=0.5, severity=0.1):
    if random.random() > probability:
        return data
    transformed = np.fft.fft2(data)
    transformed = np.fft.fftshift(transformed)
    H, W = data.shape[1:]
    mask = np.zeros((H, W))
    center = (H // 2, W // 2)
    radius = 2
    for i in range(H):
        for j in range(W):
            if (i - center[0]) ** 2 + (j - center[1]) ** 2 <= radius ** 2:
                mask[i, j] = 1
    mask = np.fft.ifftshift(mask)
    transformed *= mask
    gibbs_data = np.real(np.fft.ifft2(transformed))
    return np.clip(gibbs_data, 0, 255).astype(np.uint8)


def adjust_contrast(data, probability=0.5, factor=0.5):
    if random.random() > probability:
        return data
    mean = np.mean(data, axis=(1, 2), keepdims=True)
    contrast_data = np.clip(mean + factor * (data - mean), 0, 255)
    return contrast_data.astype(np.uint8)


def random_erasing(data, probability=0.5, max_area=0.1):
    if random.random() > probability:
        return data
    _, H, W = data.shape
    area = H * W
    target_area = random.uniform(0, max_area) * area
    aspect_ratio = random.uniform(0.3, 3.3)
    h = int(round((target_area * aspect_ratio) ** 0.5))
    w = int(round((target_area / aspect_ratio) ** 0.5))
    if h < H and w < W:
        x1 = random.randint(0, H - h)
        y1 = random.randint(0, W - w)
        data[:, x1:x1 + h, y1:y1 + w] = 0
    return data


def augment_data(data):
    data = add_gaussian_noise(data)
    data = add_gibbs_noise(data)
    data = adjust_contrast(data)
    data = random_erasing(data)
    return data


def center_crop(tensor, target_height=128, target_width=128):
    C, H, W = tensor.shape
    top = (H - target_height) // 2
    left = (W - target_width) // 2
    bottom = top + target_height
    right = left + target_width
    return tensor[:, top:bottom, left:right]


class BraTS19_H2L(Dataset):
    def __init__(self, root="", mode='train', grade="HGG",
                 split_dict_path="H2L_split.npy", transform=None, aug=False):

        load_dict = np.load(split_dict_path, allow_pickle=True).item()

        HGG_train_lists = load_dict["HGG_train_lists_new"]
        LGG_train_lists = load_dict["LGG_train_lists_new"]
        HGG_test_lists = load_dict["HGG_test_lists_new"]
        LGG_test_lists = load_dict["LGG_test_lists_new"]

        self.save_path = root + grade + "/" + mode + "/"

        if mode == 'train' and grade == "HGG":
            self.image_lists = HGG_train_lists
        elif mode == 'test' and grade == "HGG":
            self.image_lists = HGG_test_lists
        elif mode == 'train' and grade == "LGG":
            self.image_lists = LGG_train_lists
        elif mode == 'test' and grade == "LGG":
            self.image_lists = LGG_test_lists

        self.aug = aug

    def __len__(self):
        return len(self.image_lists)

    def __getitem__(self, i):
        pi = self.image_lists[i]
        img_path = self.save_path + "img_" + pi
        lab_path = self.save_path + "lab_" + pi
        pid = pi

        img = np.load(img_path)
        label = np.load(lab_path)

        if self.aug:
            img2 = img.copy()
            img2 = adjust_light(img2)
            img2 = eraser(img2)

            t1_data2 = normalize(img2[0][np.newaxis, :, :])
            t1ce_data2 = normalize(img2[1][np.newaxis, :, :])
            t2_data2 = normalize(img2[2][np.newaxis, :, :])
            flair_data2 = normalize(img2[3][np.newaxis, :, :])

            input_data2 = np.concatenate(
                (t1_data2, t1ce_data2, t2_data2, flair_data2), axis=0
            )

        t1_data = normalize(img[0][np.newaxis, :, :])
        t1ce_data = normalize(img[1][np.newaxis, :, :])
        t2_data = normalize(img[2][np.newaxis, :, :])
        flair_data = normalize(img[3][np.newaxis, :, :])

        label = re_label(label)

        input_data = np.concatenate(
            (t1_data, t1ce_data, t2_data, flair_data), axis=0
        )

        if self.aug:
            return input_data, input_data2, label, pid

        return input_data, label, pid
